### PR TITLE
Profiles: send ens flow gas

### PR DESCRIPTION
**Pull request by benisgold**
Fixes RNBW-3605

## What changed (plus any additional context for devs)
- modify gas estimation depending on send ens configs
- get rid of `shouldUseMulticall` in favor of `getTransactionTypeForRecords`, which takes into account setAddr in addition to setText

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels?
- [ ] Added e2e tests? if not please specify why
- [ ] If you added new files, did you update the CODEOWNERS file?

### CODE REVIEW

File: src/screens/SendConfirmationSheet.tsx [hunk 1]
```diff
@@ -289,6 +299,73 @@ export default function SendConfirmationSheet() {
...
+                  ...records,
+                  [recordKey]: '',
+                };
+              }, {});
+              return estimateENSSetRecordsGasLimit({
+                name: asset?.name,
+                records: emptyRecords,
+              });
+            case 'set-address':
+              return estimateENSSetNameGasLimit({
```
Comment (jxom) [inline src/screens/SendConfirmationSheet.tsx:322]: Just a note, but we don't actually call `setName` for "set-address", it's actually batched up with the set records stuff. Check out [here](https://github.com/rainbow-me/rainbow/pull/3389/files#diff-00be62de6f93f2c78dd5e9401fd2e6fb92d1f534b7b60593e012b7bb32779c0dR142-R147) and [here](https://github.com/rainbow-me/rainbow/pull/3389/files#diff-00be62de6f93f2c78dd5e9401fd2e6fb92d1f534b7b60593e012b7bb32779c0dR166-R170).

Comment (benisgold) [reply src/screens/SendConfirmationSheet.tsx:322]: cool ty

**Comment by benisgold** [comment]
I think the changes to these files may have gotten carried over from prior changes to @jxom's branch @jxom/ens-send-flow-ui. I'm guessing they should all be removed?

`android/app/src/main/java/me/rainbow/MainActivity.java-e
src/components/discover-sheet/RegisterENSSection.tsx
src/components/expanded-state/UniqueTokenExpandedState.tsx
src/hooks/useExpandedStateNavigation.js`

File: android/app/src/main/java/me/rainbow/MainActivity.java-e [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  @Override
+  protected ReactActivityDelegate createReactActivityDelegate() {
+    return new ReactActivityDelegate(this, getMainComponentName()) {
+      @Override
+      protected ReactRootView createRootView() {
+       return new RNGestureHandlerEnabledRootView(MainActivity.this);
+      }
+      @Override
+      protected void loadApp(String appKey) {
```
Comment (estebanmino) [inline android/app/src/main/java/me/rainbow/MainActivity.java-e:69]: was this file intended?

File: src/components/discover-sheet/RegisterENSSection.tsx [hunk 3]
```diff
@@ -29,13 +29,13 @@ export default function RegisterENSSection() {
...
-    } else {
-      watchingAlert();
-    }
+    navigate(Routes.REGISTER_ENS_NAVIGATOR, {
+      fromDiscover: true,
+    });
+    // if (!isReadOnlyWallet || enableActionsOnReadOnlyWallet) {
+    // } else {
+    //   watchingAlert();
+    // }
```
Comment (estebanmino) [inline src/components/discover-sheet/RegisterENSSection.tsx:38]: uncomment these?

File: src/components/expanded-state/UniqueTokenExpandedState.tsx [hunk 4]
```diff
@@ -352,11 +352,12 @@ const UniqueTokenExpandedState = ({
   const yPosition = useSharedValue(0);
 
   const profilesEnabled = useExperimentalFlag(PROFILES);
-  const isActionsEnabled = !external && !isReadOnlyWallet;
+  const isActionsEnabled = true || (!external && !isReadOnlyWallet);
   const hasSendButton = isActionsEnabled && isSendable;
 
   const hasEditButton =
-    isActionsEnabled && profilesEnabled && isENS && ensProfile.isOwner;
+    true ||
```
Comment (estebanmino) [inline src/components/expanded-state/UniqueTokenExpandedState.tsx:359]: this one feels not ready

File: src/handlers/ens.ts [hunk 5]
```diff
@@ -771,11 +802,26 @@ export const estimateENSSetRecordsGasLimit = async ({
...
-      ...{ name, ownerAddress, records: ensRegistrationRecords },
-    });
+    if (shouldUseMulticall) {
+      gasLimit = await estimateENSMulticallGasLimit({
+        name,
+        ownerAddress,
+        records: ensRegistrationRecords,
+      });
+    } else {
+      if (ensRegistrationRecords?.coinAddress?.length) {
```
Comment (estebanmino) [inline src/handlers/ens.ts:812]: nice setAddress addition 👌 

setAddress is meant to just one coin, if there are 2 coinAddress we'll be trying to send setAddr txs. So if there's more we should use multicall, maybe you could add this condition to `shouldUseMulticallTransaction` ? so if there's only one coinAddress or only one text record we should use setAddr or setText respectively 

WDYT?

File: src/handlers/ens.ts [hunk 6]
```diff
@@ -861,10 +907,9 @@ export const shouldUseMulticallTransaction = (
     text,
   } = registrationRecords;
   if (
-    !coinAddress?.length &&
     !contentHash &&
     !ensAssociatedAddress &&
-    text?.length === 1
+    (text?.length || 0) + (coinAddress?.length || 0) === 1
```
Comment (estebanmino) [inline src/handlers/ens.ts:912]: ah ok i see this condition to not use muticall when there's only one text or only one coinAddress

File: src/handlers/ens.ts (hunk 5 above)
Comment (estebanmino) [reply src/handlers/ens.ts:812]: since we have 3 types of txs now maybe we should replace `shouldUseMulticallTransaction` to a method that returns the type of tx that we're gonna use? to make it cleaner like

```
const getRecordsTypeTransaction = () => {
  if (more than one text record and/or coinAddress) return MULTICALL
  if (only one text record) return SET_TEXT
  if (only one coinAddress) return SET_ADDR
}
```
and then estimate the gas corresponding on the records tx type

File: src/hooks/useExpandedStateNavigation.js [hunk 7]
```diff
@@ -29,10 +29,10 @@ export default function useExpandedStateNavigation(inputType) {
...
   return useCallback(
     (routeName, traverseParams) => {
-      if (isReadOnlyWallet) {
-        watchingAlert();
-        return;
-      }
+      // if (isReadOnlyWallet) {
+      //   watchingAlert();
+      //   return;
+      // }
```
Comment (estebanmino) [inline src/hooks/useExpandedStateNavigation.js:35]: uncomment?

File: src/screens/SendConfirmationSheet.tsx [hunk 8]
```diff
@@ -289,6 +301,82 @@ export default function SendConfirmationSheet() {
...
+          };
+        }, {});
+        if (sendENSOptions['set-address']) {
+          records = { ...records, ETH: toAddress };
+        } else {
+          records = { ...records, ETH: AddressZero };
+        }
+        promises.push(
+          estimateENSSetRecordsGasLimit({
+            name: asset?.name,
```
Comment (estebanmino) [inline src/screens/SendConfirmationSheet.tsx:334]: i think this one could fail for ENS with non ASCII characters, if that´s the case we should clean the name before estimating gas

File: src/screens/SendConfirmationSheet.tsx [hunk 9]
```diff
@@ -289,6 +301,82 @@ export default function SendConfirmationSheet() {
...
+            records: records,
+            ownerAddress: accountAddress,
+          })
+        );
+      } else if (sendENSOptions['set-address']) {
+        promises.push(
+          estimateENSSetAddressGasLimit({
+            name: asset?.name,
```
Comment (estebanmino) [inline src/screens/SendConfirmationSheet.tsx:342]: i think this one could fail for ENS with non ASCII characters, if that´s the case we should clean the name before estimating gas

File: src/screens/SendConfirmationSheet.tsx [hunk 10]
```diff
@@ -289,6 +301,82 @@ export default function SendConfirmationSheet() {
...
+            records: formatRecordsForTransaction({ ETH: toAddress }),
+          })
+        );
+      }
+      if (sendENSOptions['transfer-control']) {
+        promises.push(
+          estimateENSSetOwnerGasLimit({
+            name: asset?.name,
```
Comment (estebanmino) [inline src/screens/SendConfirmationSheet.tsx:350]: i think this one could fail for ENS with non ASCII characters, if that´s the case we should clean the name before estimating gas

File: src/screens/SendConfirmationSheet.tsx [hunk 11]
```diff
@@ -289,6 +301,82 @@ export default function SendConfirmationSheet() {
...
+            ownerAddress: toAddress,
+            fromAddress: accountAddress,
+          })
+        );
+      }
+      promiseUtils
+        .PromiseAllWithFails(promises)
+        .then(gasLimits => {
+          const gasLimit = gasLimits.reduce(
+            (a, b) => parseFloat(a) + parseFloat(b),
```
Comment (estebanmino) [inline src/screens/SendConfirmationSheet.tsx:360]: could you use `add` from `utilities`, gasLimits are in wei units so we can have problems parsing them as floats

File: src/screens/SendConfirmationSheet.tsx [hunk 12]
```diff
@@ -645,6 +733,15 @@ export default function SendConfirmationSheet() {
               testID="send-confirmation-button"
             />
           </SendButtonWrapper>
+          {isENS && (
+            <GasSpeedButton
+              asset={null}
+              currentNetwork={network}
+              testID={null}
+              theme={isDarkMode ? 'dark' : 'light'}
+              validateGasParams={null}
```
Comment (estebanmino) [inline src/screens/SendConfirmationSheet.tsx:742]: removing these null params is not possible?

File: src/screens/SendConfirmationSheet.tsx (hunk 8 above)
Comment (estebanmino) [reply src/screens/SendConfirmationSheet.tsx:334]: just want to make it clear for people reviewing the PR (talking with ben about this internally), the problem is not the ASCII character is that opensea returns an alert icon together with the name
image

File: src/handlers/ens.ts (hunk 5 above)
Comment (benisgold) [reply src/handlers/ens.ts:812]: yeah thats a good call

File: src/screens/SendConfirmationSheet.tsx [hunk 13]
```diff
@@ -289,6 +304,81 @@ export default function SendConfirmationSheet() {
...
 
+  useEffect(() => {
+    if (isENS) {
+      const promises = [
+        estimateGasLimit({
+          address: accountAddress,
+          amount: 0,
+          asset: asset,
+          recipient: toAddress,
+        }),
```
Comment (jxom) [inline src/screens/SendConfirmationSheet.tsx:315]: This function in the `SendSheet` sets the padded gas flag https://github.com/rainbow-me/rainbow/blob/develop/src/screens/SendSheet.js#L479

should we do it here?

File: src/handlers/ens.ts [hunk 14]
```diff
@@ -785,14 +819,31 @@ export const estimateENSSetRecordsGasLimit = async ({
...
   if (validRecords) {
-    const shouldUseMulticall = shouldUseMulticallTransaction(
-      ensRegistrationRecords
-    );
-    gasLimit = await (shouldUseMulticall
-      ? estimateENSMulticallGasLimit
-      : estimateENSSetTextGasLimit)({
-      ...{ name, ownerAddress, records: ensRegistrationRecords },
-    });
+    const txType = getTransactionTypeForRecords(ensRegistrationRecords);
```
Comment (estebanmino) [inline src/handlers/ens.ts:822]: 🐐 🐐 🐐 🐐 🐐

File: src/handlers/ens.ts [hunk 15]
```diff
@@ -785,14 +819,31 @@ export const estimateENSSetRecordsGasLimit = async ({
...
+        gasLimit = await estimateENSMulticallGasLimit({
+          name,
+          ownerAddress,
+          records: ensRegistrationRecords,
+        });
+        break;
+      case ENSRegistrationTransactionType.SET_ADDR:
+        gasLimit = await estimateENSSetAddressGasLimit({
+          name,
+          records: ensRegistrationRecords,
```
Comment (estebanmino) [inline src/handlers/ens.ts:834]: `setAddr` doesn't need the `ownerAddress` param?

File: src/raps/registerENS.ts [hunk 16]
```diff
@@ -163,16 +182,29 @@ export const createTransferENSRap = async (
...
+        case ENSRegistrationTransactionType.SET_ADDR:
+          rapActionType = RapActionTypes.setAddrENS;
+          break;
+        case ENSRegistrationTransactionType.SET_TEXT:
+          rapActionType = RapActionTypes.setTextENS;
+          break;
+        default:
+          rapActionType = null;
+          break;
+      }
```
Comment (estebanmino) [inline src/raps/registerENS.ts:200]: this switch is in the code about 3 times already, could we create another helper to get the rapActionType directly?

**Review by jxom — APPROVED** [review]

**Review by estebanmino — APPROVED** [review]
